# **Model Training and Feature Engineering**

In [1]:
#### TEMPORARY
import sys
sys.path.append('/home/bac/code/kaggle/kaggle-cmi-detect-behavior/')

In [3]:
import os
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import lightgbm as lgb
from sklearn.model_selection import StratifiedGroupKFold
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import f1_score, balanced_accuracy_score

warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (12, 8)

In [4]:
from src.config import PROJECT_PATH, DATA_PATH, USE_WANDB, WANDB_PROJECT, WANDB_ENTITY
from src.tracking import ExperimentTracker

In [5]:
def display_all(df):
    with pd.option_context("display.max_rows", 100, "display.max_columns", 1000):
        display(df)

In [6]:
# Initialize the experiment tracker
tracker = ExperimentTracker(
    project_path=PROJECT_PATH,
    use_wandb=USE_WANDB,
    wandb_project_name=WANDB_PROJECT,
    wandb_entity=WANDB_ENTITY
)

In [8]:
# Load data
train_sensor = pd.read_csv(os.path.join(os.path.expanduser(DATA_PATH), 'train.csv'))
train_demos = pd.read_csv(os.path.join(os.path.expanduser(DATA_PATH), 'train_demographics.csv'))

# Merge
train_df = pd.merge(train_sensor, train_demos, on='subject', how='left')
display_all(train_df.head())

row_id sequence_type sequence_id  sequence_counter      subject  \
0  SEQ_000007_000000        Target  SEQ_000007                 0  SUBJ_059520   
1  SEQ_000007_000001        Target  SEQ_000007                 1  SUBJ_059520   
2  SEQ_000007_000002        Target  SEQ_000007                 2  SUBJ_059520   
3  SEQ_000007_000003        Target  SEQ_000007                 3  SUBJ_059520   
4  SEQ_000007_000004        Target  SEQ_000007                 4  SUBJ_059520   

                       orientation                                   behavior  \
0  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
1  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
2  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
3  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
4  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   

        phase             gesture     acc_x     acc_y     acc_z     rot_w  \
0  Transition  Cheek - pinch skin  6.683594  6.214844  3.355469  0.134399   
1  Transition  Cheek - pinch skin  6.949219  6.214844  3.125000  0.143494   
2  Transition  Cheek - pinch skin  5.722656  5.410156  5.421875  0.219055   
3  Transition  Cheek - pinch skin  6.601562  3.531250  6.457031  0.297546   
4  Transition  Cheek - pinch skin  5.566406  0.277344  9.632812  0.333557   

      rot_x     rot_y     rot_z      thm_1      thm_2      thm_3      thm_4  \
0 -0.355164 -0.447327 -0.809753  28.943842  31.822186  29.553024  28.592863   
1 -0.340271 -0.428650 -0.824524  29.340816  31.874645  29.791740  28.663383   
2 -0.274231 -0.356934 -0.865662  30.339359  30.935045  30.090014  28.796087   
3 -0.264160 -0.238159 -0.885986  30.543730  27.044001  29.310717  29.018711   
4 -0.218628 -0.063538 -0.914856  29.317265  25.270855  26.808746  29.408604   

       thm_5  tof_1_v0  tof_1_v1  tof_1_v2  tof_1_v3  tof_1_v4  tof_1_v5  \
0  28.310535     131.0     134.0     132.0     135.0      98.0      74.0   
1  28.406172     130.0     138.0     131.0     135.0     101.0      76.0   
2  28.529778     137.0     136.0     147.0     109.0      90.0      81.0   
3  27.402010     143.0     147.0     170.0     127.0     109.0      98.0   
4  27.357603     178.0     191.0     183.0     157.0     146.0     139.0   

   tof_1_v6  tof_1_v7  tof_1_v8  tof_1_v9  tof_1_v10  tof_1_v11  tof_1_v12  \
0      64.0      60.0      -1.0      -1.0      152.0      153.0      141.0   
1      66.0      61.0      -1.0      -1.0      156.0      155.0      141.0   
2      74.0      74.0      -1.0     164.0      165.0      146.0      106.0   
3      95.0      95.0      -1.0     177.0      189.0      177.0      136.0   
4     143.0     148.0      -1.0      -1.0      236.0      238.0      208.0   

   tof_1_v13  tof_1_v14  tof_1_v15  tof_1_v16  tof_1_v17  tof_1_v18  \
0       89.0       68.0       63.0       -1.0       -1.0       -1.0   
1       93.0       74.0       64.0       -1.0       -1.0       -1.0   
2       94.0       77.0       77.0       -1.0       -1.0       -1.0   
3      121.0      107.0      104.0       -1.0       -1.0       -1.0   
4      200.0      185.0      190.0       -1.0       -1.0       -1.0   

   tof_1_v19  tof_1_v20  tof_1_v21  tof_1_v22  tof_1_v23  tof_1_v24  \
0       -1.0      169.0      118.0       86.0       73.0       -1.0   
1       -1.0      165.0      116.0       86.0       75.0      130.0   
2      180.0      140.0      118.0      103.0       92.0       -1.0   
3      202.0      171.0      160.0      141.0      135.0       -1.0   
4      210.0      246.0      225.0      228.0      202.0      149.0   

   tof_1_v25  tof_1_v26  tof_1_v27  tof_1_v28  tof_1_v29  tof_1_v30  \
0       -1.0       -1.0       -1.0       -1.0      147.0      110.0   
1       -1.0       -1.0       -1.0       -1.0      142.0      114.0   
2       -1.0       -1.0       -1.0       -1.0      155.0      119.0   
3       -1.0       -1.0       -1.0    

In [9]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 574945 entries, 0 to 574944
Columns: 348 entries, row_id to elbow_to_wrist_cm
dtypes: float64(334), int64(6), object(8)
memory usage: 1.5+ GB


### Feature Engineering and Training - Wave 0 - Baseline

In [10]:
def create_baseline_features(df):
    """
    Create Wave 0 features: Simple sequence level aggregations.
    """
    aggs = {
        'acc_x': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'acc_y': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'acc_z': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'rot_w': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'rot_x': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'rot_y': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'rot_z': ['mean', 'std', 'min', 'max', 'median', 'skew'],
    }
    
    # Add Thermopile aggregations
    for i in range(1, 6):
        aggs[f'thm_{i}'] = ['mean', 'std', 'min', 'max']
        
    # Add time of flight aggregations (simple mean and std for baseline)
    tof_cols = [f'tof_{s}_v{p}' for s in range(1, 6) for p in range(64)]
    # Replace -1 with NaN to correctly calculate stats
    df[tof_cols] = df[tof_cols].replace(-1, np.nan)
    aggs['tof_mean'] = ['mean', 'std']
    df['tof_mean'] = df[tof_cols].mean(axis=1)
    
    # Group by sequence and aggregate and flatten multi index cols
    agg_df = df.groupby('sequence_id').agg(aggs)
    agg_df.columns = ['_'.join(col).strip() for col in agg_df.columns.values]
    
    # Get sequence level metadata (target, subject, etc.)
    meta_df = df.groupby('sequence_id').first()
    
    # Combine aggregated features with metadata
    final_df = pd.concat([meta_df[['subject', 'gesture']], agg_df], axis=1)
    
    # Encode gesture target
    final_df['gesture_encoded'] = final_df['gesture'].astype('category').cat.codes
    
    print(f"Feature engineering complete. Shape of features: {final_df.shape}")
    return final_df

In [11]:
# Create features
features_df = create_baseline_features(train_df)
display_all(features_df.head())

Feature engineering complete. Shape of features: (8151, 67)


subject                   gesture  acc_x_mean  acc_x_std  \
sequence_id                                                                 
SEQ_000007   SUBJ_059520        Cheek - pinch skin    6.153098   1.334155   
SEQ_000008   SUBJ_020948  Forehead - pull hairline    3.400506   1.087142   
SEQ_000013   SUBJ_040282        Cheek - pinch skin   -7.058962   1.295184   
SEQ_000016   SUBJ_052342         Write name on leg    5.524654   1.074108   
SEQ_000018   SUBJ_032165  Forehead - pull hairline    5.363715   1.627637   

             acc_x_min  acc_x_max  acc_x_median  acc_x_skew  acc_y_mean  \
sequence_id                                                               
SEQ_000007    3.613281   9.015625      6.488281   -0.545319    3.915570   
SEQ_000008    1.734375   5.906250      3.437500    0.146452    5.311179   
SEQ_000013   -9.250000  -3.347656     -7.144531    0.518519    2.346182   
SEQ_000016    3.437500   9.378906      5.390625    0.747648   -4.408491   
SEQ_000018    1.964844   6.832031      6.101562   -1.397824    4.109737   

             acc_y_std  acc_y_min  acc_y_max  acc_y_median  acc_y_skew  \
sequence_id                                                              
SEQ_000007    3.048287  -2.019531   6.519531      5.488281   -1.184870   
SEQ_000008    3.268073  -0.222656   8.667969      7.000000   -0.830467   
SEQ_000013    2.564639  -3.273438   4.683594      3.382812   -1.445762   
SEQ_000016    0.598318  -5.718750  -2.960938     -4.492188    0.505319   
SEQ_000018    3.525304  -3.164062   6.718750      6.007812   -1.347944   

             acc_z_mean  acc_z_std  acc_z_min  acc_z_max  acc_z_median  \
sequence_id                                                              
SEQ_000007     5.577782   2.337517   1.093750   9.792969      4.964844   
SEQ_000008     6.581629   2.475402   1.722656  11.074219      5.839844   
SEQ_000013    -6.068544   1.330784 -10.945312  -3.515625     -5.851562   
SEQ_000016    -3.162077   6.139752  -8.078125   8.355469     -6.667969   
SEQ_000018     5.937066   2.104544   4.148438   9.933594      4.761719   

             acc_z_skew  rot_w_mean  rot_w_std  rot_w_min  rot_w_max  \
sequence_id                                                            
SEQ_000007     0.586111    0.263574   0.069033   0.134399   0.379272   
SEQ_000008     0.186346    0.243493   0.064414   0.157593   0.341980   
SEQ_000013    -1.039566    0.392208   0.150629   0.061157   0.540771   
SEQ_000016     0.964846    0.361083   0.041568   0.277527   0.459045   
SEQ_000018     1.040372    0.859159   0.034238   0.828247   0.925049   

             rot_w_median  rot_w_skew  rot_x_mean  rot_x_std  rot_x_min  \
sequence_id                                                               
SEQ_000007       0.254578    0.307637   -0.280817   0.056597  -0.442871   
SEQ_000008       0.226562    0.326730   -0.117145   0.049384  -0.263306   
SEQ_000013       0.439514   -1.359968    0.340804   0.182002   0.140991   
SEQ_000016       0.352234    0.525728   -0.728107   0.207529  -0.893677   
SEQ_000018       0.846283    1.030080    0.177468   0.178091  -0.184204   

             rot_x_max  rot_x_median  rot_x_skew  rot_y_mean  rot_y_std  \
sequence_id                                                               
SEQ_000007   -0.204163     -0.275757   -0.846226   -0.331470   0.175050   
SEQ_000008   -0.050537     -0.097382   -0.937947   -0.342327   0.190164   
SEQ_000013    0.726501      0.258362    1.350803    0.800506   0.090017   
SEQ_000016   -0.384827     -0.857361    0.843064   -0.223281   0.156706   
SEQ_000018    0.305542      0.270111   -1.342766   -0.352176   0.149264   

             rot_y_min  rot_y_max  rot_y_median  rot_y_skew  rot_z_mean  \
sequence_id                                                               
SEQ_000007   -0.478027   0.005066     -0.414978    1.258707   -0.837994   
SEQ_000008   -0.508606  -0.031555     -0.442169    0.830673   -0.875143   
SEQ_000013    0.580505   0.881653      0.838135   -1.593

In [12]:
#Config
EXPERIMENT_NAME = "Baseline-Wave0-LGBM-2"
MODEL_NAME = "LightGBM"
FEATURE_WAVE = "Wave-0"
N_SPLITS = 5
SEED = 42

# Model Params
params = {
    'objective': 'multiclass',
    'num_class': features_df['gesture_encoded'].nunique(),
    'metric': 'multi_logloss',
    'n_estimators': 1000,
    'learning_rate': 0.05,
    'feature_fraction': 0.8,
    'bagging_fraction': 0.8,
    'bagging_freq': 1,
    'lambda_l1': 0.1,
    'lambda_l2': 0.1,
    'num_leaves': 31,
    'verbose': -1,
    'n_jobs': -1,
    'seed': SEED,
    'boosting_type': 'gbdt',
}

# Prep data for CV
X = features_df.drop(columns=['subject', 'gesture', 'gesture_encoded'])
y = features_df['gesture_encoded']
groups = features_df['subject']

In [13]:
# Cross Validation Setup
cv = StratifiedGroupKFold(n_splits=N_SPLITS, shuffle=True, random_state=SEED)

oof_preds = np.zeros((len(features_df), y.nunique()))
oof_true = np.zeros(len(features_df))
fold_scores = []

# Training loop
for fold, (train_idx, val_idx) in enumerate(cv.split(X, y, groups)):
    print(f"--- Fold {fold+1}/{N_SPLITS} ---")
    
    # Split data
    X_train, y_train = X.iloc[train_idx], y.iloc[train_idx]
    X_val, y_val = X.iloc[val_idx], y.iloc[val_idx]
    
    # Init and train model
    model = lgb.LGBMClassifier(**params)
    model.fit(X_train, y_train,
              eval_set=[(X_val, y_val)],
              eval_metric='multi_logloss',
              callbacks=[lgb.early_stopping(100, verbose=False)])
    
    # Predict on validation set
    val_preds = model.predict_proba(X_val)
    oof_preds[val_idx] = val_preds
    oof_true[val_idx] = y_val
    
    # Eval fold performance using balanced_accuracy_score since it is robust to class imbalance
    fold_score = balanced_accuracy_score(y_val, np.argmax(val_preds, axis=1))
    fold_scores.append(fold_score)
    print(f"Fold {fold+1} Balanced Accuracy: {fold_score:.5f}")
    
# Final score and logging
mean_cv_score = np.mean(fold_scores)
print("\n--- CV Summary ---")
print(f"Mean Balanced Accuracy: {mean_cv_score:.5f}")
print(f"Std Dev.: {np.std(fold_scores):.5f}")

# Log the experiment
tracker.log_experiment(
    experiment_name=EXPERIMENT_NAME,
    model_name=MODEL_NAME,
    feature_wave=FEATURE_WAVE,
    cv_score=mean_cv_score,
    params=params,
    notes="Inital baseline model with sequence-level statistical aggregates."
)    

--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.60245
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.52807
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.54332
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.48495
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.52315

--- CV Summary ---
Mean Balanced Accuracy: 0.53639
Std Dev.: 0.03821
Experiment 'Baseline-Wave0-LGBM-2' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


wandb: Currently logged in as: b-a-chaudhry (b-a-chaudhry-) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


Experiment 'Baseline-Wave0-LGBM-2' logged to W&B


#### Analysis:

1. With 18 classes, a random classifier would achieve a balanced accuracy of approximately 1/18 (~5.6%). The current score of ~54% is nearly 10 times better and is a strong confirmation that even the simplest `Wave-0` features contain a strong predictive signal.
2. The standard deviation of 0.0381 across the folds is relatively low, indicating that the model's performance is consistent and not wildly dependent on a specific subset of subjects. 
3. The benchmark score ofo 0.53639 is now the number to beat! 